### PR TITLE
Enable Precise Location

### PR DESCRIPTION
**Pull request by malmstein**
Task/Issue URL: https://app.asana.com/0/488551667048375/1174485623962657
Tech Design URL: https://app.asana.com/0/488551667048375/1178719929289442
Tech Design URL: https://app.asana.com/0/488551667048375/1179723338484874

**Description**:
Our Android app does not currently support precise user location, because the app does not have location permission. This PR enables this feature for all domains that support it.

The flow and explanation of Custom Dialog 1 / System permission dialog / Custom Dialog 2 ca be found [here](https://app.asana.com/0/488551667048375/1184034713968714)

**Steps to test this PR**:

***Fresh Install scenario***:
Location Sharing Disabled
1. Open Phone / Settings
2. Disable Location Sharing
3. Open app
4. Visit http://html5demos.com/geo
5. Location is denied

***All scenarios below assume that Location Sharing is enabled***

Deny Custom Dialog 1 
1. Open app
2. Visit http://html5demos.com/geo
3. Custom Dialog 1 appears
4. Click "Maybe Later"
5. Pixel "m_pc_syd_l" is sent
6. Location is denied
7. Close tab
8. Open new tab and visit Visit http://html5demos.com/geo
9. Custom Dialog 1 appears again

Deny Site Custom Dialog 1 
1. Open app
2. Visit http://html5demos.com/geo
3. Custom Dialog 1 appears
4. Click "Deny always for this site"
6. Pixel "m_pc_syd_n" is sent
7. Pixel "m_pc_sd_da" is sent
8. Location is denied
9. Close tab
10. Open new tab and visit Visit http://html5demos.com/geo
11. Location is denied

Deny system permission
1. Open app
2. Visit http://html5demos.com/geo
3. Custom Dialog 1 appears
4. Click Enable
5. Pixel "m_pc_syd_e" is sent
6. System dialog permission appears
7. Deny permission
8. Location is denied

Allow Always permission
1. Open app
2. Visit http://html5demos.com/geo
3. Custom Dialog 1 appears
4. Click Enable
5. Pixel "m_pc_syd_e" is sent
6. System dialog permission appears
7. Accept permission
8. Pixel "m_pc_s_l_e" is sent
9. Custom Dialog 2 appears
10. Accept Always
11. Pixel "m_pc_sd_aa" is sent
12. Location is now found
13. Close tab
14. Open new tab and visit Visit http://html5demos.com/geo
15. Location should be found

Allow this Session permission
1. Open app
2. Visit http://html5demos.com/geo
3. Custom Dialog 1 appears
4. Accept permission
6. Pixel "m_pc_syd_e" is sent
7. System dialog permission appears
8. Accept permission
9. Pixel "m_pc_s_l_e" is sent
10. Custom Dialog 2 appears
11. Accept this session
12. Pixel "m_pc_sd_ao" is sent
13. Location is now found
14. Close tab
15. Open new tab and visit Visit http://html5demos.com/geo
16. Custom Dialog 2 should appear again

Deny always permission
1. Open app
2. Visit http://html5demos.com/geo
3. Custom Dialog 1 appears
4. Accept permission
5. Pixel "m_pc_syd_e" is sent
6. System dialog permission appears
7. Accept permission
8. Pixel "m_pc_s_l_e" is sent
9. Custom Dialog 2 appears
10. Deny Always
11. Pixel "m_pc_sd_da" is sent
12. Location is denied
13. Close tab
14. Open new tab and visit Visit http://html5demos.com/geo
15. Location should be denied

Deny this Session permission
1. Open app
2. Visit http://html5demos.com/geo
3. Custom Dialog 1 appears
4. Accept permission
5. Pixel "m_pc_syd_e" is sent
6. System dialog permission appears
7. Accept permission
8. Pixel "m_pc_s_l_e" is sent
9. Custom Dialog 2 appears
10. Deny this session
11. Pixel "m_pc_sd_do" is sent
12. Location is denied
13. Close tab
14. Open new tab and visit Visit http://html5demos.com/geo
15. Custom Dialog 2 should appear again

***System permission already granted***:
Same scenarios as with fresh install, but Custom Dialog 1 and System Dialog permission should not appear

***Interaction with Settings screen***:
Edit permissions from Settings
1. Open app
2. Visit http://html5demos.com/geo
3. Custom Dialog 2 appears
4. Deny always
5. Location is denied
6. Open Settings
7. Tap on Overflow icon
8. Edit Permission
9. Choose Enable
10. Go back to http://html5demos.com/geo
11. Location should be shared

Edit permissions from Settings
1. Open app
2. Visit http://html5demos.com/geo
3. Custom Dialog 2 appears
4. Allow always
5. Location is shared
6. Open Settings
7. Tap on Overflow icon
8. Edit Permission
9. Choose Disable
10. Go back to http://html5demos.com/geo
11. Location is denied

Remove permissions from Settings
1. Open app
2. Visit http://html5demos.com/geo
3. Custom Dialog 2 appears
4. Allow always
5. Location is shared
6. Open Settings
7. Tap on Overflow icon
8. Delete Permission
9. Go back to http://html5demos.com/geo
10. Custom Dialog 2 should appear

Permissions toggle in Settings / Location
1. Open app
2. Visit https://duckduckgo.github.io/privacy-test-pages/features/geolocation.html
3. Allow permission Always
5. Location is shared
6. Open Settings / Location
7. Disable
8. List of permissions should be empty
9. Visit https://duckduckgo.github.io/privacy-test-pages/features/geolocation.html
10. Custom Dialog 2 should appear

***Sites should not be able to refresh location in the background***:
Tab in background (assuming system location permission has been granted)
1. Open app
2. Visit https://duckduckgo.github.io/privacy-test-pages/features/geolocation.html
3. Custom Dialog 2 appears
4. Allow always
5. A new location update should appear, more or less 2s ago
6. Open a new tab
7. Navigate to another site for about 30 seconds or so
8. Navigate back to the initial tab
9. A new location should appear with 30s ago update

App in background (assuming system location permission has been granted)
1. Open app
2. Visit https://duckduckgo.github.io/privacy-test-pages/features/geolocation.html
3. Custom Dialog 2 appears
4. Allow always
5. A new location update should appear, more or less 2s ago
6. Put the app in background
7. Wait for 30 seconds or so
8. Navigate back to the app
9. A new location should appear with 30s ago update

***Interaction with Fire Button**:
(Configure Fire Button to Clear Data and Tabs, only Tabs won't affect the behavior)
1. Open app
2. Visit https://duckduckgo.github.io/privacy-test-pages/features/geolocation.html
3. Custom Dialog 2 appears
4. Allow always
5. Use Fire Button
6. Open Settings / Location 
7. List should be empty
8. Visit https://duckduckgo.github.io/privacy-test-pages/features/geolocation.html
7. Custom Dialog should appear again

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)

### CODE REVIEW

**Comment by CDRussell** [comment]
@malmstein, there's something wrong with the latest commits. Now, on a fresh install, `checkSystemLocationPermission` is finding `PERMISSION_DENIED` (as expected) but is treating that like the user hit "NEVER ALLOW".

The logic for handling `ActivityCompat.shouldShowRequestPermissionRationale` is inversed; if `shouldShowRequestPermissionRationale` is true, then we know the user hit "NEVER". otherwise, we should show the dialog as that means the permission hasn't been granted _yet_

**Comment by CDRussell** [comment]
I can't get location working on either html5demos or privacy test pages site when running on a Lollipop device. Have you tested it on Lollipop? Anything special I have to do here?

File: app/src/main/java/com/duckduckgo/app/global/view/StringHtmlExtension.kt [hunk 1]
```diff
@@ -35,3 +35,16 @@ private fun htmlDrawable(context: Context, resource: Int): Drawable {
     drawable.setBounds(0, 0, drawable.intrinsicWidth, drawable.intrinsicHeight)
     return drawable
 }
+
+private const val HTTPS_PREFIX = "https://"
+private const val WWW_PREFIX = "https://www."
+private const val WWW_SUFFIX = "/"
+
+fun String.website(): String {
+    return this.takeIf { it.startsWith(WWW_PREFIX, ignoreCase = true) && it.endsWith(WWW_SUFFIX, ignoreCase = true) }
```
Comment (CDRussell) [inline app/src/main/java/com/duckduckgo/app/global/view/StringHtmlExtension.kt:44]: I'm confused what this function is doing. It drops `https://www` from the start, and a trailing `/` from the end, but only if both are present? `&&`

what about `https://www.example.com`, would we want to drop the prefix from that too even though it doesn't have a trailing slash? 

(or maybe i'm not understanding the function fully; i find it hard to parse)

File: app/src/main/java/com/duckduckgo/app/location/GeoLocationPermissionsManager.kt [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    override fun clear(domain: String) {
+        val geolocationPermissions = GeolocationPermissions.getInstance()
+        geolocationPermissions.clear(domain)
+    }
+
+    override suspend fun clearAll() {
+        withContext(dispatchers.io()) {
+            val geolocationPermissions = GeolocationPermissions.getInstance()
+            val permissions = permissionsRepository.getLocationPermissionsSync()
+            val fireproofed = fireproofWebsiteRepository.getFireproofWebsitesSync()
```
Comment (CDRussell) [inline app/src/main/java/com/duckduckgo/app/location/GeoLocationPermissionsManager.kt:55]: We don't appear to do anything with this variable other than logging. is it still worth doing the lookup just for that?

File: app/src/main/java/com/duckduckgo/app/location/data/LocationPermissionEntity.kt [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+@Entity(tableName = "locationPermissions")
+data class LocationPermissionEntity(
+    @PrimaryKey val domain: String,
+    val permission: LocationPermissionType
+)
+
+private const val HTTPS_HOST_PREFIX = "https://"
+private const val WWW_SUFFIX = "/"
+
+fun LocationPermissionEntity.forFireproofing(): String {
```
Comment (CDRussell) [inline app/src/main/java/com/duckduckgo/app/location/data/LocationPermissionEntity.kt:31]: This seems almost identical to the function in `StringHtmlExtension`. Worth combining?

File: app/src/main/java/com/duckduckgo/app/location/ui/LocationPermissionsAdapter.kt [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+
+    fun updatePermissions(
+        isLocationPermissionEnabled: Boolean,
+        newPermissions: List<LocationPermissionEntity>
+    ) {
+        locationPermissionEnabled = isLocationPermissionEnabled
+        locationPermissions = newPermissions
+        notifyDataSetChanged()
+
+        // val diffCallback = DiffCallback(locationPermissions, newPermissions)
```
Comment (CDRussell) [inline app/src/main/java/com/duckduckgo/app/location/ui/LocationPermissionsAdapter.kt:66]: 🔥

File: app/src/main/java/com/duckduckgo/app/location/ui/LocationPermissionsAdapter.kt [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+                    if (it.permission == LocationPermissionType.ALLOW_ONCE || it.permission == LocationPermissionType.ALLOW_ALWAYS) {
+                        allowedLocationPermissions.add(it)
+                    } else {
+                        deniedLocationPermissions.add(it)
+                    }
+                }
+            }
+        }
+
+    private var allowedLocationPermissions: MutableList<LocationPermissionEntity> = mutableListOf()
```
Comment (CDRussell) [inline app/src/main/java/com/duckduckgo/app/location/ui/LocationPermissionsAdapter.kt:90]: It feels strange having these member variables declared throughout the file, and not all at the top

File: app/src/main/java/com/duckduckgo/app/location/ui/SiteLocationPermissionDialog.kt [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+        hideExtraViews(allowOnce, denyOnce, extraDivider, anotherDivider)
+        makeCancellable()
+
+        return alertDialog.create()
+    }
+
+    private fun populateTitle(title: TextView) {
+        arguments?.let { args ->
+            val originUrl = args.getString(KEY_REQUEST_ORIGIN)
+            val dialogTitle = getString(R.string.preciseLocationSiteDialogTitle, originUrl.website())
```
Comment (CDRussell) [inline app/src/main/java/com/duckduckgo/app/location/ui/SiteLocationPermissionDialog.kt:84]: `nullable` string lint warning here

File: app/src/main/java/com/duckduckgo/app/location/ui/SiteLocationPermissionDialog.kt [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+                dismiss()
+                listener.onSiteLocationPermissionSelected(originUrl, LocationPermissionType.ALLOW_ONCE)
+            }
+            denyOnce.setOnClickListener {
+                dismiss()
+                listener.onSiteLocationPermissionSelected(originUrl, LocationPermissionType.DENY_ONCE)
+            }
+            denyAlways.setOnClickListener {
+                dismiss()
+                listener.onSiteLocationPermissionSelected(originUrl, LocationPermissionType.DENY_ALWAYS)
```
Comment (CDRussell) [inline app/src/main/java/com/duckduckgo/app/location/ui/SiteLocationPermissionDialog.kt:123]: lint warnings on nullability of `originUrl`. If we're confident it'll never be null, we should make that more explicit. With the warnings, it isn't clear if we're sure or just missed the warning

File: app/src/main/java/com/duckduckgo/app/location/ui/SystemLocationPermissionDialog.kt [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+        if (arguments == null) throw IllegalArgumentException("Missing arguments bundle")
+        val args = requireArguments()
+        if (!args.containsKey(KEY_REQUEST_ORIGIN)) {
+            throw IllegalArgumentException("Bundle arguments required [KEY_REQUEST_ORIGIN")
+        }
+    }
+
+    private fun populateSubtitle(view: TextView) {
+        arguments?.let { args ->
+            val originUrl = args.getString(KEY_REQUEST_ORIGIN).website()
```
Comment (CDRussell) [inline app/src/main/java/com/duckduckgo/app/location/ui/SystemLocationPermissionDialog.kt:75]: `nullable` warning

File: app/src/androidTest/java/com/duckduckgo/app/browser/BrowserTabViewModelTest.kt [hunk 9]
```diff
@@ -2310,9 +2326,339 @@ class BrowserTabViewModelTest {
     }
 
     @Test
-    fun whenDosAttackDetectedThenErrorIsShown() {
```
[the remark this answers is not in the document]
Comment (malmstein) [reply app/src/androidTest/java/com/duckduckgo/app/browser/BrowserTabViewModelTest.kt:2313]: Merge issue, fixing now!

File: app/src/main/java/com/duckduckgo/app/browser/BrowserTabFragment.kt [hunk 10]
```diff
@@ -97,24 +145,51 @@ import com.duckduckgo.app.widget.ui.AddWidgetInstructionsActivity
 import com.duckduckgo.widget.SearchWidgetLight
 import com.google.android.material.snackbar.Snackbar
 import dagger.android.support.AndroidSupportInjection
-import kotlinx.android.synthetic.main.fragment_browser_tab.*
-import kotlinx.android.synthetic.main.include_add_widget_instruction_buttons.view.*
-import kotlinx.android.synthetic.main.include_cta_buttons.view.*
-import kotlinx.android.synthetic.main.include_dax_dialog_cta.*
-import kotlinx.android.synthetic.main.include_find_in_page.*
-import kotlinx.android.synthetic.main.include_new_browser_tab.*
+import kotlinx.android.synthetic.main.fragment_browser_tab.autoCompleteSuggestionsList
```
[the remark this answers is not in the document]
Comment (malmstein) [reply app/src/main/java/com/duckduckgo/app/browser/BrowserTabFragment.kt:148]: I've removed any formatting from my side now. I was worried it would be `ktlint` formatting but we explicitly exclude `no-wildcard-imports` from being formatted.

File: app/src/main/java/com/duckduckgo/app/browser/BrowserTabFragment.kt [hunk 11]
```diff
@@ -599,6 +701,15 @@ class BrowserTabFragment : Fragment(), FindListener, CoroutineScope, DaxDialogLi
         }
     }
 
+    private fun showMessageSnackbar(domain: String) {
+        val snackbar = Snackbar.make(rootView, getString(R.string.preciseLocationSnackbarMessage, domain.website()), Snackbar.LENGTH_SHORT)
+        val snackbarView = snackbar.view
```
[the remark this answers is not in the document]
Comment (malmstein) [reply app/src/main/java/com/duckduckgo/app/browser/BrowserTabFragment.kt:706]: wasn't sure but I think it makes sense

File: app/src/main/java/com/duckduckgo/app/browser/BrowserTabViewModel.kt [hunk 12]
```diff
@@ -278,6 +291,9 @@ class BrowserTabViewModel(
     val title: String?
         get() = site?.title
 
+    private lateinit var permissionOrigin: String
+    private lateinit var permissionCallback: GeolocationPermissions.Callback
```
[the remark this answers is not in the document]
Comment (malmstein) [reply app/src/main/java/com/duckduckgo/app/browser/BrowserTabViewModel.kt:295]: This won't work because it's a property and must be initialized. We could initialize them with `emptyString` and a `FakeCallback` or initialize them as null. Making them nullable forces us to always check for nullability, which is the same we have to do now and we wouldn't be adding any improvement. Do you see a better option?

File: app/src/main/java/com/duckduckgo/app/browser/BrowserTabViewModel.kt [hunk 13]
```diff
@@ -730,6 +749,24 @@ class BrowserTabViewModel(
...
 
+    private suspend fun notifyPermanentLocationPermission(domain: String) {
+        if (!appSettingsPreferencesStore.appLocationPermission) {
+            return
+        }
+
+        Timber.i("Precise Location - Checking for $domain Location Permission")
+        val userHasGivenPermission = locationPermissionsRepository.hasUserGivenPermissionTo(domain)
+        if (userHasGivenPermission) {
+            val permissionEntity = locationPermissionsRepository.getDomainPermission(domain)!!
```
[the remark this answers is not in the document]
Comment (malmstein) [reply app/src/main/java/com/duckduckgo/app/browser/BrowserTabViewModel.kt:760]: Yes, I think that's cleaner, I tried to hide the `nullable` but it's really not our API to fix.

File: app/src/main/java/com/duckduckgo/app/browser/BrowserTabViewModel.kt [hunk 14]
```diff
@@ -801,6 +838,140 @@ class BrowserTabViewModel(
...
+            onSiteLocationPermissionAlwaysDenied()
+            return
+        }
+
+        if (origin != null) {
+            viewModelScope.launch {
+                val userHasGivenPermission = locationPermissionsRepository.hasUserGivenPermissionTo(permissionOrigin)
+                val previouslyDeniedForever = appSettingsPreferencesStore.appLocationPermissionDeniedForever
+                if (userHasGivenPermission) {
+                    val permissionEntity = locationPermissionsRepository.getDomainPermission(permissionOrigin)!!
```
[the remark this answers is not in the document]
Comment (malmstein) [reply app/src/main/java/com/duckduckgo/app/browser/BrowserTabViewModel.kt:860]: ditto

File: app/src/main/java/com/duckduckgo/app/browser/BrowserChromeClient.kt [hunk 15]
```diff
@@ -120,4 +121,9 @@ class BrowserChromeClient @Inject constructor(private val uncaughtExceptionRepos
     override fun onCloseWindow(window: WebView?) {
         webViewClientListener?.closeCurrentTab()
     }
+
+    override fun onGeolocationPermissionsShowPrompt(origin: String?, callback: GeolocationPermissions.Callback?) {
```
[the remark this answers is not in the document]
Comment (malmstein) [reply app/src/main/java/com/duckduckgo/app/browser/BrowserChromeClient.kt:125]: you are right, not sure what happened there but this definitely saves up a few lines

File: app/src/main/java/com/duckduckgo/app/location/GeoLocationPermissionsManager.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (malmstein) [reply app/src/main/java/com/duckduckgo/app/location/GeoLocationPermissionsManager.kt]: yep, can be removed

File: app/src/main/java/com/duckduckgo/app/global/view/StringHtmlExtension.kt (hunk 1 above)
Comment (malmstein) [reply app/src/main/java/com/duckduckgo/app/global/view/StringHtmlExtension.kt:44]: Agree naming could be better. This is formatting the `origin` that we receive from `onGeolocationPermissionsShowPrompt`. The format is always the same, `https://www` (http is no longer allowed) and `/` at the end.

How about something like `String.websiteFromGeoLocationsApiOrigin()`?

File: app/src/main/java/com/duckduckgo/app/location/data/LocationPermissionEntity.kt (hunk 3 above)
Comment (malmstein) [reply app/src/main/java/com/duckduckgo/app/location/data/LocationPermissionEntity.kt:31]: not the same though, `LocationPermissionEntity` domains don't have `www` (because the fireproofed websites are stored like that) but the ones coming from the `GeolocationAPI` do. So I wanted to keep them separated

File: app/src/main/java/com/duckduckgo/app/location/ui/LocationPermissionsAdapter.kt (hunk 5 above)
Comment (malmstein) [reply app/src/main/java/com/duckduckgo/app/location/ui/LocationPermissionsAdapter.kt:85]: it does feel strange, I'm moving them to the proper place

File: app/src/main/java/com/duckduckgo/app/location/ui/SiteLocationPermissionDialog.kt (hunk 6 above)
Comment (malmstein) [reply app/src/main/java/com/duckduckgo/app/location/ui/SiteLocationPermissionDialog.kt:84]: we know it can't be `null` as it's part of the `instance` call so we are safe to use `!!` here

File: app/src/main/java/com/duckduckgo/app/location/ui/SiteLocationPermissionDialog.kt (hunk 7 above)
Comment (malmstein) [reply app/src/main/java/com/duckduckgo/app/location/ui/SiteLocationPermissionDialog.kt:123]: yeah, as described above we are safe to make it explicit.

File: app/src/main/java/com/duckduckgo/app/browser/BrowserTabViewModel.kt (hunk 12 above)
[the remark this answers is not in the document]
Comment (CDRussell) [reply app/src/main/java/com/duckduckgo/app/browser/BrowserTabViewModel.kt:295]: >Making them nullable forces us to always check for nullability, which is the same we have to do now 

I think the way we do it now is wrong, in that because it's lateinit, it'll never be null. lint is telling us that.
Screen Shot 2020-08-14 at 5 11 18 PM

That null check isn't actually keeping us safe. if we access that variable and it wasn't initialized, the app will crash. I think initializing to `nullable` is the least likely to bite us later

File: app/src/main/java/com/duckduckgo/app/global/view/StringHtmlExtension.kt (hunk 1 above)
Comment (CDRussell) [reply app/src/main/java/com/duckduckgo/app/global/view/StringHtmlExtension.kt:44]: Thanks for explaining. I think it's partly the _fancy_ kotlin that's throwing me here. You could take advantage of the fact that the suffix isn't case-sensitive, and write it in a more imperative way that i think is easier to understand e.g.
```
fun String.websiteFromGeoLocationsApiOrigin(): String {
    var extracted = this.removeSuffix(WWW_SUFFIX)

    if (extracted.startsWith(WWW_PREFIX, ignoreCase = true)) {
        extracted = extracted.drop(WWW_PREFIX.length)
    }

    return extracted
}
```

But two things

1. Might just be a personal preference thing; will leave it with you
1. I haven't tested that code! 😅

File: app/src/main/java/com/duckduckgo/app/browser/BrowserTabViewModel.kt (hunk 12 above)
[the remark this answers is not in the document]
Comment (CDRussell) [reply app/src/main/java/com/duckduckgo/app/browser/BrowserTabViewModel.kt:296]: Any update on this query?

File: app/src/main/java/com/duckduckgo/app/browser/BrowserTabFragment.kt [hunk 16]
```diff
@@ -584,6 +646,33 @@ class BrowserTabFragment : Fragment(), FindListener, CoroutineScope, DaxDialogLi
         }
     }
 
+    private fun checkSystemLocationPermission(domain: String, deniedForever: Boolean) {
+        if (ContextCompat.checkSelfPermission(requireActivity(), Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
```
Comment (CDRussell) [inline app/src/main/java/com/duckduckgo/app/browser/BrowserTabFragment.kt:650]: We ask for both `ACCESS_FINE_LOCATION` and `ACCESS_COARSE_LOCATION`, but only check the result for `ACCESS_FINE_LOCATION`. Does this matter?

**Comment by CDRussell** [comment]
Almost good to go! One last thing, `clearAllButFireproofed` isn't working for me; it's clearing out fireproofed sites too.

**Comment by malmstein** [comment]
Thanks for the review @CDRussell!
